src/main/java/org/verdictdb/jdbc41/VerdictStatement.java

```diff
@@ -35,11 +43,82 @@
 
   VerdictSingleResult result;
 
+  private VerdictDBLogger log = VerdictDBLogger.getLogger(this.getClass());
+
   public VerdictStatement(Connection conn, VerdictContext context) {
     this.conn = conn;
     this.executionContext = context.createNewExecutionContext();
   }
 
+  public ResultSet sql(String query) throws VerdictDBException {
+    VerdictStreamResultSet resultSet = new VerdictStreamResultSet();
+    if (getStreamQueryIfExists(query).getKey()) {
+      query = query.replaceFirst("(?i)stream", "");
+      VerdictResultStream resultStream = executionContext.streamsql(query);
+      new Thread(new ExecuteStream(resultStream, resultSet)).start();
+      return resultSet;
+    }
+    return null;
+  }
+
+  /**
+   * Created by Shucheng Zhong on 9/13/18
+   * <p>
+   * It will try to get the VerdictSingleResult from VerdictResultStream
+   * and append the VerdictSingleResult to VerdictStreamResultSet
+   * until all the VerdictSingleResults have return.
+   * Since it running on separate thread, it won't block user querying from
+   * the ResultSet
+   */
+  class ExecuteStream implements Runnable {
+
+    VerdictResultStream resultStream;
+
+    VerdictStreamResultSet resultSet;
+
+    ExecuteStream(VerdictResultStream resultStream, VerdictStreamResultSet resultSet) {
+      this.resultStream = resultStream;
+      this.resultSet = resultSet;
+    }
+
+    /**
+     * When the last SingleResult is returned, it will synchronized the isCompleted flag of VerdictStreamResultSet
+     * to block possible call of next(). It will set status of VerdictStreamResultSet to be complete and
+     * append the last SingleResult to VerdictStreamResultSet.
+     */
+    public void run() {
+      while (!resultStream.isCompleted()) {
+        while (resultStream.hasNext()) {
+          VerdictSingleResult singleResult = resultStream.next();
+          if (!resultStream.hasNext()) {
+            synchronized ((Object) resultSet.isCompleted) {
```

Comment:
Can you elaborate in the above comment about the purpose of this synchronized block? What could happen without the existence of this block?

Reply:
I think the reason why I added this block is because VerdictStreamResultSet also check isCompleted flag in next() method. I need to make this block atomic to prevent that VerdictStreamResultSet is unaware that the process has ended after the last result is appended. I will add this to comment.